qmetaobject/src/log.rs

```diff
@@ -0,0 +1,218 @@
+//! Logging facilities and forwarding
+
+use std::os::raw::c_char;
+
+use log::{Level, logger, Record};
+
+use crate::QString;
+
+use self::QtMsgType::*;
+
+cpp! {{
+    #include <qmetaobject_rust.hpp>
+}}
+
+cpp_class!(
+    /// Wrapper for Qt's QMessageLogContext
+    pub unsafe struct QMessageLogContext as "QMessageLogContext"
+);
+
+impl QMessageLogContext {
+    // Return QMessageLogContext::line
+    pub fn line(&self) -> i32 {
+        cpp!(unsafe [self as "QMessageLogContext*"] -> i32 as "int" { return self->line; })
+    }
+    // Return QMessageLogContext::file
+    pub fn file(&self) -> &str {
+        unsafe {
+            let x = cpp!([self as "QMessageLogContext*"] -> *const c_char as "const char*" {
+                return self->file;
+            });
+            if x.is_null() {
+                return "";
+            }
+            std::ffi::CStr::from_ptr(x).to_str().unwrap()
+        }
+    }
+    // Return QMessageLogContext::function
+    pub fn function(&self) -> &str {
+        unsafe {
+            let x = cpp!([self as "QMessageLogContext*"] -> *const c_char as "const char*" {
+                return self->function;
+            });
+            if x.is_null() {
+                return "";
+            }
+            std::ffi::CStr::from_ptr(x).to_str().unwrap()
+        }
+    }
+    // Return QMessageLogContext::category
+    pub fn category(&self) -> &str {
+        unsafe {
+            let x = cpp!([self as "QMessageLogContext*"] -> *const c_char as "const char*" {
+                return self->category;
+            });
+            if x.is_null() {
+                return "";
+            }
+            std::ffi::CStr::from_ptr(x).to_str().unwrap()
+        }
+    }
+}
+
+/// Wrap Qt's QtMsgType enum
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum QtMsgType {
+    QtDebugMsg,
+    QtWarningMsg,
+    QtCriticalMsg,
+    QtFatalMsg,
+    QtInfoMsg,
+    // there is also one level defined in C++ code:
+    // QtSystemMsg = QtCriticalMsg
+}
+
+/// Mapping from Qt logging levels to Rust logging facade's levels.
+///
+/// Due to the limited range of levels from both sides,
+/// [`QtCriticalMsg`][`QtMsgType`] and [`QtFatalMsg`][`QtMsgType`]
+/// both map to [`log::Level::Error`][Level],
+/// while [`log::Level::Trace`][Level] is never returned.
+///
+/// [Level]: https://docs.rs/log/0.4.10/log/enum.Level.html
+/// [`QtMsgType`]: https://doc.qt.io/qt-5/qtglobal.html#QtMsgType-enum
+pub fn map_level(lvl: QtMsgType) -> Level {
```

Comment:
I wouldn't put these two function `pub`
The From  impls should be enough.

Reply:
OK, reverted back to `From`.